keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,766 @@
+
+In addition, users may expect that the order of resource deletions inside a namespace is guaranteed even when requested through the namespace deletion.
+For alpha, resources with liens are deleted in a nondeterministic order on the namespace deletion.
+Handling order of deletion is a beta blocker and won't be included in the initial implementation.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, the API server must block any request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp`.
+
+#### Race of removing/adding liens
+
+A cluster-admin would be racing to update the object to remove the lien and delete the object before a namespace editor is able to place the lien back. As a result, a namespace admin or editor can create an object that a cluster-admin or namespace lifecycle controller cannot delete, which shouldn't be allowed.
+
+To mitigate the risk, a new `IgnoreLiens` API option in `DeleteOptions` to force delete a resource with liens will be added.
+
+Specifying the option should only be allowed to a limited set of users and groups, therefore deletion request will be blocked in api-server, if invalid users request deletion with the option.
+Restricting the `IgnoreLiens` API option to specific users is beta blocker and won't be included in the initial implementation.
+Therefore, in the initial implementation, any users with delete permission will be able to delete resources with liens by specifying the `IgnoreLiens` API option.
+
+#### Subresource
+
+There is a use case to disallow a controller to write to its CRD except for the status subresource, but allow it to update finalizers/liens.
+To allow this use case, it is required to provide a feature to set a separate RBAC to update finalizers/liens.
+On the other hand, finalizers field is currently not subresourced.
+Therefore, it needs to provide this feature in a compatible way for finalizers, and it would be better to implement liens field in the same way to finalizers.
+
+To solve this issue, there is [another KEP](https://github.com/kubernetes/enhancements/issues/3285) to add a new knob for just patching only a specific field via an endpoint of a subresource, without modifying the field itself.
+In this way, a feature to set a separate RBAC can be implemented later without making liens field itself a subresource.
+
+Subresourcing `Liens` is a beta blocker and won't be included in the initial implementation.
```

Comment:
I think this needs to be resolved for alpha so that our clients will function properly across versions.  Similar comment here: https://github.com/kubernetes/enhancements/pull/2840#discussion_r898167946

Reply:
My take: I'm assuming that mutating liens will be allowed through update/patch as well as a subresource (just like annotations and labels).  So a client written to use PATCH will still work, and when we add the subresource it will be *possible* to convert (but not *required*).

David - are you imagining otherwise?  If the ONLY way to mutate liens is thru the subresource (e.g. fail any update/patch which changes them -- which has a certain appeal) then I agree it is a requirement.

Reply:
>  If the ONLY way to mutate liens is thru the subresource (e.g. fail any update/patch which changes them -- which has a certain appeal) 

I had been thinking of it in terms of "only the liens subresource modifies liens".  I can see a use-case for setting liens on creation, but for updates where the object has already existed without liens for some time, I don't see a reason why separate REST calls are particularly bad.

I can probably be persuaded that the CREATE and spec UPDATE are congruent.  But I see status update rules as more like what we built for CRs, where the *only* fields allowed to be updated are under status.

I don't see a way to build what I would consider a safe, general liens handler without a subresource and I think that sort of client handling is important to get right initially.

Reply:
> I had been thinking of it in terms of "only the liens subresource modifies liens". 

Do we have any other examples?

Labels, annotations, and finalizers (all in metadata) can be updated by PATCH/PUT.

Scale (in spec) can be updated by PATCH/PUT *or* the subresource.

Status is the only one I know that is different.

I'm not against it - it's just different.

Also see https://github.com/kubernetes/enhancements/pull/2840#discussion_r900286510 - let's assume that the subresource is the ONLY way to update lines.  Alice is a namespace user - she can create Pods.  If a lien gets put on my pod - do we want Alice to be able to clear it?

Reply:
> I had been thinking of it in terms of "only the liens subresource modifies liens"

that's not great for human beings adding the lien themselves, does it?